Archival sync in batches & parallel.

goral09 opened the pull request:
Closes https://github.com/casper-network/casper-node/issues/3015

The fast-sync algorithm consists of three phases:
1. Initial sync - where we download the trusted block and the trie (state) under its state root hash.
2. syncing to Genesis (if `archival_sync=true`) - where we download all data from before the trusted block: all blocks, deploys and all of the state tries under all blocks.
3. syncing forward - where we synchronize chain from the trusted block onwards, to the latest block at the tip.

This PR affects only the second step in that process.

Historically (before this PR), the _sync to Genesis_ part would start at trusted block and then sequentially download its parent block, its deploys and the state trie for that block. That process would be repeated until we hit Genesis.

This turned out to be too slow - the requested/response for block, independent requests for all of its deploys and then multiple, independent requests for the trie took much longer than expected. The process was estimated to finish (for mainnet data) in *years*.

This PR changes the algorithm (of **2. Syncing to Genesis only**) in the following way:
1. Starting from trusted block, fetch batch of headers (currently a batch of 1000) from `trusted_block.height..=trusted_block.height - 1000`. Every batch is validated (for continuity, proper parent-child relationship, cryptographic verification).
2. Continue syncing batches until we hit Genesis. 
3. After we hit Genesis, we start "walking back" - from Genesis to the trusted block. 
4. We spawn up to `max_parallel_block_fetches` workers, each fetching block AND its deploy (another change compared to the original algorithm, this is safe b/c block and deploys' hashes are not allowed to be bigger than network message size limit). After block with deploys is downloaded, we start downloading its state - the trie under its state root hash. We spawn up to `max_parallel_trie_fetches` workers (shared across all "block workers") to synchronize the trie.
5. When "block workers" recognize that the next block they would be downloading is the trusted block (or its child- by height) - they stop and exit.
6. When all block workers finish - we're synchronized the archival state leading up to the trusted block.

Commit:
Fetch headers to Genesis in batches.

Commit:
Parallelize block and state downloads.

Commit:
Log warning if we received fetch response for no responder.

Commit:
Fix BatchId iterator and add some basic tests.

Commit:
Fix batch response id calculation in fetcher.

Commit:
More tests for BlockHeadersBatchId

Commit:
Test BlockHeadersBatch::is_continuous_and_descending

Commit:
Unit test for BlockHeadersBatch::from_vec

Commit:
Add unit tests for BlockHeadersBatch::validate

Commit:
Disconnect from a sender that sends INVALID data

Commit:
Replace panics with proper errors

Commit:
lints

Commit:
Remove old fetch_to_genesis code

goral09 commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1074,71 +1075,169 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+            Ok(fetched_block) => {
+                trace!(?block_hash, "downloaded block and deploys");
+                // We want to download the trie only when we know we have the block.
+                sync_trie_store(*fetched_block.block.state_root_hash(), ctx).await?;
```
`sync_trie_store_worker` we will create `max_parallel_trie_fetches` meaning that the overall process of `fetch_blocks_and_state_since_genesis` will end up creating up to `max_parallel_block_fetches * max_parallel_trie_fetches` parallel `sync_trie_store_worker`s. 

One solution to that is to share the trie `WorkQueue` across all `fetch_block_worker`s.

marc-casperlabs replied:
The issue here is that the `WorkQueue` has a specific notion of "done" - if all jobs are finished and they have not added any new jobs to the queue, the entire task --- fetching all tries for a specific root hash --- is complete. This sort of signaling needs to be preserved somehow, we currently use the end of all workers as the indicator of jobs done.

marc-casperlabs replied:
A quick fix would be to use a semaphore, acquiring a ticket before spawning a worker and releasing it once a worker finishes. I'm sure we can build something better by refactoring, but that's a reasonable start.

We may need to think about persistent worker pools in the future, and properly wrap jobs being passed to them. The harder part is just signaling that a specific subset of tasks is done.

goral09 replied:
This is done in [`51a4c1e` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/51a4c1efc029fd4adeb52355bba51646026a0cd6)

Commit:
Fix batch size to 1024.

Commit:
Revert "Log warning if we received fetch response for no responder."

This reverts commit b21eda9bc694667021be2cfc104a4e2ffa06998f.

Commit:
Do not retry download when fetch_retry_forever returns an error.

Commit:
Validate batches from storage.

Commit:
Fix after bad rebase.

marc-casperlabs commented on node/src/components/chain_synchronizer/error.rs:
```diff
@@ -53,6 +54,9 @@ pub(crate) enum Error {
     #[error("no such block hash: {bogus_block_hash}")]
     NoSuchBlockHash { bogus_block_hash: BlockHash },
 
+    #[error("no such block height: {0} when archival sync")]
```
Maybe something like "no such block height: {0} encountered during archival sync"?

Fraser999 replied:
We need to use the term "sync to genesis" rather than "archival sync"

goral09 replied:
Will do.

marc-casperlabs commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1074,71 +1075,182 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+                lowest_trusted_block_header = new_lowest;
+            }
+            Err(err) => {
+                // If we get an error here it means something must have gone really wrong.
```
I dig the long explanatory comment :+1:

marc-casperlabs commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1074,71 +1075,182 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+            .await
+        {
+            Some(hash) => hash,
+            None => return Err(Error::NoSuchBlockHeight(block_height)),
```
If this is error is only used here, the error name and message and such could reflect the fact that we expected the block to be there, since its header was fetched.

goral09 replied:
You're right. I was thinking that I accomodate for any future use of the error but I should focus on the task at hand and expect devs coming after me to update the error message if necessary.

marc-casperlabs commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1074,71 +1075,182 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+    // Flag set by a worker when it encounters an error.
+    let abort = Arc::new(AtomicBool::new(false));
+
+    for block_height in 0..=ctx.config.max_parallel_block_fetches() as u64 {
```
Are you sure this is the correct range? I believe it should fetch up to the latest block, not `ctx.config.max_parallel_block_fetches()`?

goral09 replied:
This only creates a correct number of workers. The workers themselves enqueue next block to download.

marc-casperlabs commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1074,71 +1075,182 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+
+    while let Some(result) = workers.next().await {
+        result?; // Return the error if a download failed.
+        ctx.metrics.chain_sync_blocks_synced.inc();
```
This will record the number of workers that finished, not the number of blocks synced. IIRC workers should exit only when there are no more jobs coming in, and each worker should usually downloading more than a single block?

goral09 replied:
Good catch. Thanks.

goral09 replied:
Done in [`8bf9239` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/8bf923978268ac8fca4d75b89066772fa0b65dab)

marc-casperlabs commented on node/src/components/fetcher.rs:
```diff
@@ -603,6 +603,42 @@ impl ItemFetcher<BlockHeader> for Fetcher<BlockHeader> {
 }
 
+impl ItemFetcher<BlockHeadersBatch> for Fetcher<BlockHeadersBatch> {
+    const SAFE_TO_RESPOND_TO_ALL: bool = false;
```
I don't think it matters, but I wonder if this is actually `true` :).

goral09 replied:
You're probably right since there is only one valid batch - there's only one valid block at any block height. I'll change.

goral09 replied:
Done in [`9c7c3e2` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/9c7c3e2c3492f1f6f3d6583055ec6d6c33cfe53b)

Commit:
Add unit test for BlockHeadersBatch::validate

marc-casperlabs commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,209 @@ impl Item for BlockHeaderWithMetadata {
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+/// ID idenfitifying a request for a batch of block headers.
+pub struct BlockHeadersBatchId {
+    pub highest: u64,
```
I think the block header batches should be hardcoded, with the height either being fixed or determined by the chainspec. For example, if that number is 1024, batch `0` is always going to be headers `genesis..1023`, batch `1` blocks `1024-2047` and so on. I think that would obviate a lot of checking.

goral09 replied:
Maybe, that's a different approach than what I took. I'll stick to what we have since it's complete.

marc-casperlabs commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,209 @@ impl Item for BlockHeaderWithMetadata {
+            return Err(BlockHeadersBatchValidationError::BatchNotContinuous);
+        }
+
+        self.0
```
This could also include a size check. If we check anything but the latest chunk, we could indicate we expect this to be of a specific length.

goral09 replied:
Yeah, we can do that.

goral09 replied:
Done in [`9a8a3cc` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/9a8a3ccb0d005f93b230be0a166628b293b7b903)

marc-casperlabs commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,209 @@ impl Item for BlockHeaderWithMetadata {
+        batch: &[BlockHeader],
+        verifiable_chunked_hash_activation: EraId,
+    ) -> bool {
+        batch
```
While I do appreciate functional iterator magic, I think 

```rust
for i in 1..batch.len() {
    let current = batch[i];
    let pred = batch[i-1];

     if current.parent_hash() != pred.hash(...) { ... }
     if curent.height() != pred.height() +1 { ... }
}
```

is maybe a little easier to read and review? :)

goral09 replied:
I actually find the "iterator magic" more declarative:
* create windows of size 2
* filter out any window that doesn't have 2 elements
* validate pairs

goral09 replied:
I'm going to resolve the comment for now :)

Commit:
Verify length of BlockHeadersBatch response.

Commit:
Update block downloaded metric correctly.

Commit:
It's safe to respond to all for BlockHeadersBatch

Commit:
Push block height with hash for better progress tracking.

Commit:
Downgrade logs from INFO to TRACE

Commit:
Downgrade bad peer logs to debug

xcthulhu commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+        self.0
+            .last()
+            .cloned()
+            .ok_or(BlockHeadersBatchValidationError::BatchEmpty)
```
Why is this returning a block header instead of `()`?

goral09 replied:
When we sync headers in batches, back to Genesis, we always start with some "known/trusted" header and then calculate the batch we want to get from there. We fetch these batches sequentially, the first batch starts with _trusted header_, the next one should start at header with height `trusted_header.height - BATCH_SIZE`. After fetching the first batch, we want to get the lowest, currently known header. If we didn't return it from this method we would have to read it from storage.

afck commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1074,71 +1075,189 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
-                if lowest_available_block.block.height() == 0 {
-                    ctx.effect_builder
-                        .update_lowest_available_block_height_in_storage(0)
+        if lowest_trusted_block_header.height() == 0 {
```
Why not just make it a `while` loop?

afck commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1074,71 +1075,189 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+            .await
+        {
+            Some(hash) => hash,
+            None => return Err(Error::NoSuchBlockHeight(block_height)),
```
What if the current chain length is less than the number of workers?

Wouldn't it be simpler to just use a `height_counter: AtomicU64` instead of the work queue? The workers would, in a loop, `let height = height_counter.fetch_add(1, SeqCst);`, read the hash from storage and download block, deploys and global state. It would exit the loop if `height >= ctx.trusted_block_header().height()`.

That way the number of workers would occur only in one place, and not inside the worker logic.

goral09 replied:
> What if the current chain length is less than the number of workers?

Good point. 

I also like your suggestion for alternative. I'll see to integrate it.

goral09 replied:
Done in [`fb0e801` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/fb0e801c87c60a207cbb44d59cdc021ad78e2ac3).

afck commented on node/src/components/fetcher.rs:
```diff
@@ -603,6 +603,42 @@ impl ItemFetcher<BlockHeader> for Fetcher<BlockHeader> {
+
+    fn peer_timeout(&self) -> Duration {
+        self.get_from_peer_timeout
+    }
```
(I wonder whether this trait is a bit redundant: Would it be simpler to move these three getters into `impl<T> Fetcher<T>`, and `get_from_storage` to `Item`? Anyway, has nothing to do with this PR.)

afck commented on node/src/types/error.rs:
```diff
@@ -142,3 +142,15 @@ pub(crate) enum BlockWithMetadataValidationError {
+    BatchNotContinuous,
+    #[error("Empty batch")]
+    BatchEmpty,
+    #[error("Hash of the highest block from batch doesn't match expected. Expected: {expected}, got: {got}.")]
```
(Line too long.)

goral09 replied:
You're killing me with this Andreas :) We shouldn't be expected to manually format things that rustfmt can't.

afck commented on node/src/types/error.rs:
```diff
@@ -142,3 +142,15 @@ pub(crate) enum BlockWithMetadataValidationError {
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub(crate) enum BlockHeadersBatchValidationError {
+    #[error("Batch has incorrect lenght. Expect: {expected}, got: {got}")]
```
```suggestion
    #[error("Batch has incorrect length. Expected: {expected}, got: {got}")]
```

goral09 replied:
Seems like I got rid of this enum along the way. Resolving the comment.

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
     }
 }
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+/// ID idenfitifying a request for a batch of block headers.
```
```suggestion
/// ID identifying a request for a batch of block headers.
```

goral09 replied:
Done in [`60ac7e8` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/60ac7e8598c652d5e8d043c149df8319bdd03027)

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+impl BlockHeadersBatchId {
+    pub fn new(highest: u64, lowest: u64) -> Self {
+        Self { highest, lowest }
+    }
+    pub fn from_known(lowest_known_block_header: &BlockHeader, max_batch_size: u64) -> Self {
```
```suggestion
    }

    pub fn from_known(lowest_known_block_header: &BlockHeader, max_batch_size: u64) -> Self {
```

goral09 replied:
Done in [`60ac7e8` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/60ac7e8598c652d5e8d043c149df8319bdd03027)

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+        (self.lowest..=self.highest).rev()
+    }
+
+    /// Returns a length of the batch.
```
```suggestion
    /// Returns the length of the batch.
```

goral09 replied:
Done in [`60ac7e8` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/60ac7e8598c652d5e8d043c149df8319bdd03027)

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+
+    /// Returns a length of the batch.
+    pub fn len(&self) -> u64 {
+        self.highest + 1 - self.lowest
```
This can overflow unless `highest >= lowest && highest != u64::MAX`.

goral09 replied:
Overflow if block height is `u64::MAX`, right? We would probably have a lot more problems if that happened.

afck replied:
Right, otherwise it's an underflow.

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+    pub(crate) fn validate(
+        &self,
+        batch_id: &BlockHeadersBatchId,
+        latest_known: &BlockHeader,
```
Isn't this rather `earliest_known` or `lowest_known`?

goral09 replied:
I would say it's all of that - depending how you interpret it. Do you prefer `lowest_known`? I'll change it.

afck replied:
In what sense is it the latest known? In reverse order, because that's the order in which we're downloading them?
(I'm fine with it, if you find that more intuitive.)

goral09 replied:
since we're fetching and validating headers in descending orther, the _latest_ one we had downloaded is also _earliest_ & _lowest_.

goral09 replied:
I'll rename to `earliest_known`.

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+        match batch.first() {
+            Some(highest) => {
+                if highest.height() != requested_id.highest {
+                    error!(expected_highest=?requested_id.highest, got_highest=?highest, "unexpected highest block header");
```
(Line too long. Also below.)

goral09 replied:
Done in [`60ac7e8` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/60ac7e8598c652d5e8d043c149df8319bdd03027)

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+            .filter_map(|window| match &window {
+                &[l, r] => Some((l, r)),
+                _ => None,
+            })
```
You can use `Itertools::tuple_windows` to avoid this `filter_map` step.

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+    }
+
+    #[cfg(test)]
+    // Test-only constructor allowing creation otherwise invalid data.
```
```suggestion
    // Test-only constructor allowing creation of otherwise invalid data.
```

goral09 replied:
Done in [`60ac7e8` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/60ac7e8598c652d5e8d043c149df8319bdd03027)

afck commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+        BlockHeadersBatchId::new(
+            upper_batch_height.unwrap_or_default(),
+            lower_batch_height.unwrap_or_default(),
+        )
```
I wonder whether the batch's highest block's hash should just be part of the ID? That way, if we're somehow on the wrong chain, the other nodes wouldn't all send us back a whole range of headers that we then throw away due to the mismatch? It would also remove some of the possibly overflowing arithmetic if the ID was just the highest hash and the length.

goral09 replied:
I assumed we would recognize much earlier in the process whether we are on a wrong chain.

afck replied:
It depends. If some of our peers are trying to fool us, and we also got the wrong trusted hash?

goral09 replied:
If they try to fool us I don't think this would change much - i.e. they could ignore the request ID and respond with anything they want. 

We do validate the response against the lowest trusted block anyway so we would detect it early on whether the response if valid or not.

afck replied:
What if you're connected to 20 peers and one of them wants to fool you, and also managed (via some other channel) give you a wrong trusted hash: You'll reject all the answers from the good peers, and only accept his. And then you start requesting batches, by height, and receive ~10 full batches from good peers on average before you ask the bad peer who will finally send you the (wrong) batch that you're expecting.

If you made your requests by hash, the good peers wouldn't have to bother to respond, and could save their bandwidth. Not sure how much of a difference that would make. :shrug:

afck commented on node/src/components/storage.rs:
```diff
@@ -1186,31 +1200,41 @@ impl StorageInner {
+                        }
+                        Err(err) => {
+                            error!(?err, ?block_header_hash, "error when storing block header");
+                            txn.abort();
```
If we abort, we probably should update the block header index and remove the ones that were added during this transaction? (Also in other places where we call `put_block_header`.)

goral09 replied:
Good observation. I will change the logic to update the index only if the whole batch is stored successfuly.

goral09 replied:
Done in [`7a00eda` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/7a00edae7ec6e8fdc9853c7fd8663b94bb1201a3)

Commit:
Simplify block workers

Commit:
Log when fetch block worker is done.

Commit:
Log when sync-trie worker is done.

Commit:
Base block worker on AtomicU64 instead of using WorkQueue.

afck commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1210,16 +1195,30 @@ async fn fetch_blocks_and_state_since_genesis(ctx: &ChainSyncContext<'_>) -> Res
+        {
+            Some(hash) => hash,
+            None => return Err(Error::NoSuchBlockHeight(next_block_height)),
+        };
```
(You could use `ok_or_else` or `ok_or`.)

goral09 replied:
Of course, I will. Thanks.

afck reviewed (commented):
Thanks, the `AtomicU64` approach looks good!

Commit:
Merge remote-tracking branch 'upstream/dev' into batch-headers-sync-to-genesis

Commit:
Remove very talkative log

Commit:
Update in-memory indices only if headers were stored succesfully.

Commit:
Use the max_parallel_trie_limit globally across multiple block workers.

Commit:
Use ok_or_else instead of explicit match and return.

Commit:
Clippy.

Commit:
Log progress every 1000th block to avoid spamming

Commit:
Add a comment explaining why we have to resync every time from scratch.

Commit:
Downgrade to TRACE syncing trie store log

Commit:
Track time (and count) spent waiting on storage sempahore limiter.

Commit:
`get_trie` metric starts with smaller values.

Before started at 10ms and tracked until 10s.
Now starts at 1ms and trakcs until 1s.

Commit:
Fix storage tests.

Commit:
Fix clippy

Commit:
Various updates to formatting and spelling fixes.

Commit:
Merge remote-tracking branch 'upstream/dev' into batch-headers-sync-to-genesis

Commit:
Commit lost clippy change.

Commit:
Update config defaults

Commit:
Remove accidentally committed file.

Commit:
Merge remote-tracking branch 'upstream/dev' into batch-headers-sync-to-genesis

Fraser999 commented on utils/dry-run-deploys/Cargo.toml:
```diff
@@ -22,6 +22,7 @@ once_cell = "1.5.2"
 rand = "0.8.3"
 reqwest = { version = "0.11.1", features = ["json"] }
 retrieve-state = { path = "../retrieve-state" }
+prometheus = "0.12.0"
```
```suggestion
```

Fraser999 commented on utils/retrieve-state/Cargo.toml:
```diff
@@ -20,6 +20,7 @@ jsonrpc-lite = "0.5.0"
 lmdb = "0.8.0"
 num-rational = { version = "0.4.0", features = ["serde"] }
 reqwest = { version = "0.11.1", features = ["json", "gzip"] }
+prometheus = "0.12.0"
```
```suggestion
```

goral09 replied:
Done in [`c9cb619` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/c9cb6196ec98dcc8f347231080172be11d18479f)

Commit:
Merge remote-tracking branch 'upstream/dev' into batch-headers-sync-to-genesis

Fraser999 commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,226 @@ impl Item for BlockHeaderWithMetadata {
+                }
+            }
+            None => {
+                error!("response cannot be an empty batch");
```
Seems like the error logs in this method should be warnings only, since they could be invoked by a malicious peer sending invalid data.

goral09 replied:
Done in [`36ded88` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/36ded88067816ec81e9ed1af5e997cf372fb127c)

Fraser999 commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,226 @@ impl Item for BlockHeaderWithMetadata {
+            })
+            .all(|(l, r)| {
+                l.height() == r.height() + 1
+                    && l.parent_hash() == &r.hash(verifiable_chunked_hash_activation)
```
`r.hash` is a pretty expensive call, and we seem to be calling `is_continuous_and_descending` in a few places.

Maybe we should look to replicate the lazy-initialized `is_valid` field of `Deploy` with a `is_continuous_and_descending` field?

goral09 replied:
I'm not a fan of the approach taken with `is_valid`. I'll try to rewrite the code so that it's more efficient.

goral09 replied:
I removed calls to `is_continuous_and_descending ` from all the places except for `Item for BlockHeadersBatch` which is used in the fetcher.

Fraser999 commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,226 @@ impl Item for BlockHeaderWithMetadata {
+        if lower_batch_height.is_none() || upper_batch_height.is_none() {
+            // ID should be infallible but it is possible that the `Vec` is empty.
+            // In that case we log an error to indicate something went really wrong and use `(0,0)`.
+            error!(
```
Probably just a `warn!` here again for the same reason as above?

goral09 replied:
Done in [`36ded88` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/36ded88067816ec81e9ed1af5e997cf372fb127c)

Fraser999 commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -1084,75 +1098,191 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+            fetch_retry_forever::<BlockHeadersBatch>(ctx, batch_id).await?;
+        match fetched_headers_data {
+            FetchedData::FromStorage { item } => {
+                return BlockHeadersBatch::validate(
+                    &*item,
+                    &batch_id,
+                    lowest_trusted_block_header,
+                    ctx.config.verifiable_chunked_hash_activation(),
+                )
+                .map_err(FetchBlockHeadersBatchError::InvalidBatchFromStorage)
```
Seems like we're calling `validate` here only to get the lowest block header from the batch.  I'd be inclined to change the validation to return `Result<(), ...>` and have a separate getter on the `BlockHeadersBatch` for the lowest header.

Also, nitpick: this would more normally be written `item.validate(...)` rather than `BlockHeadersBatch::validate(&*item, ...)`.

goral09 replied:
> Also, nitpick: this would more normally be written item.validate(...) rather than BlockHeadersBatch::validate(&*item, ...).

I agree but we have the same method name (`validate`) on `Item` trait and it was resolving to that one instead of `BlockHeadersBatch`.

goral09 replied:
>Seems like we're calling validate here only to get the lowest block header from the batch. 

Not really, I also wanted to validate the batch from the storage. Although, if we agree that we should trust data from the DB I will remove that and simply get the lowest element.

goral09 replied:
Done in [`f46bbd2` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/f46bbd27aa8bf8d72a06a0b483d09e29ddc84132)

Fraser999 commented on node/src/types/block.rs:
```diff
@@ -1111,6 +1113,226 @@ impl Item for BlockHeaderWithMetadata {
 }
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+/// ID identifying a request for a batch of block headers.
+pub struct BlockHeadersBatchId {
```
I think all the new `pub` structs in this file and their fields and methods can be `pub(crate)` instead.

goral09 replied:
Done in [`36ded88` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/36ded88067816ec81e9ed1af5e997cf372fb127c)

Fraser999 reviewed (commented):
Mostly nitpicks, but I'd like feedback on the `is_continuous_and_descending` comment before approving.

Commit:
Fix error log message.

Commit:
Remove prometheus dependency.

Commit:
Change logging to warn and pub to pub(crate)

Commit:
Do not validate batch from storage.

Commit:
Do not validate batch continuity when fetched from storage.

Commit:
Don't verify continuity in  - it's already done in the fetcher.

Commit:
Remove obsolete failure case.

We don't validate continuity in the BlockHeadersBatch::validate anymore.
Continuity is validated in the Fetcher now.

goral09 commented:
bors r+

TomVasile commented:
bors try

TomVasile commented:
bors try-

TomVasile committed:
sync nctl config files

goral09 committed:
Merge pull request #2 from TomVasile/zh-sre-505

goral09 commented:
bors r+
